Create a ConfigureMainRenderTargetEvent for enabling stenciling

FiniteReality opened the pull request:
This needed a small change to mod loading: it has been separated into two separate methods instead of the one `ClientModLoader.begin` to enable configuring the render target.

The API I've implemented here is somewhat a placeholder, but ideally we would be able to add additional methods for things like additional color attachments, or potentially listeners for when the render target gets resized.

FiniteReality committed:
Separate ClientModLoader begin/finish into two methods, implement MainTarget configuration event

FiniteReality committed:
Flow stencil information through post chains

embeddedt commented on patches/net/minecraft/client/renderer/PostPass.java.patch:
```diff
@@ -0,0 +1,28 @@
+ 
+     @OnlyIn(Dist.CLIENT)
+-    public static record TargetInput(String samplerName, ResourceLocation targetId, boolean depthBuffer, boolean bilinear) implements PostPass.Input {
++    public static record TargetInput(String samplerName, ResourceLocation targetId, boolean depthBuffer, boolean bilinear, boolean stencilBuffer) implements PostPass.Input {
```
Instead of two booleans I think it makes more sense to make an enum for the type of buffer?

FiniteReality replied:
It might, but I figured that keeping the constructor MOSTLY the same would be better for compatibility

embeddedt replied:
It shouldn't matter as you added the old constructor back anyway. The only other thing you'd need to add is a `boolean depthBuffer()` getter.

embeddedt commented on patches/net/minecraft/client/renderer/PostChain.java.patch:
```diff
@@ -0,0 +1,48 @@
+             RenderTargetDescriptor rendertargetdescriptor = switch (entry.getValue()) {
+                 case PostChainConfig.FixedSizedTarget(int i, int j) -> {
+-                    yield new RenderTargetDescriptor(i, j, true);
++                    yield new RenderTargetDescriptor(i, j, useDepth, useStencil);
```
Since vanilla always enables depth should we do the same?

FiniteReality replied:
Since shaders targeting vanilla only work on the main render target, which has depth enabled, this should effectively be the same - It should only be different if somebody tries to set up a render target with depth disabled which I feel is unlikely to happen anyway

shartte commented on src/main/java/net/neoforged/neoforge/client/event/ConfigureMainRenderTargetEvent.java:
```diff
@@ -0,0 +1,72 @@
+ * This event is fired on the mod-speciffic event bus, only on the {@linkplain LogicalSide#CLIENT logical client}.
+ */
+public class ConfigureMainRenderTargetEvent extends Event implements IModBusEvent {
+    private boolean useDepth;
```
I'd remove this if it cannot be set to false (and why would it be possible to do so)

FiniteReality replied:
The reason why I kept it is because vanilla doesn't exactly use the depth buffer themselves*, so they *may* change it to false in The Future:tm: for VRAM purposes. Furthermore, it allows mods to be explicit about whether they need the depth information to be written. (I see as this a good thing)

The defaults are basically lifted from vanilla code currently, though it may be better to inline them where the event is actually constructed

shartte replied:
I'd not keep this "in reserve" in the API. If Vanilla does add it, we can re-add it then.

Shadows-of-Fire replied:
This functionality should be removed for now.  We can re-assess if Mojang changes something or there is a formal feature request to allow it.

FiniteReality replied:
Instead of removing it completely, I instead changed the patch to pass this boolean, as well as the desired width and height to the event. As we always pass `true`, I made it final and removed the set/enable method.

shartte commented on patches/com/mojang/blaze3d/pipeline/RenderTarget.java.patch:
```diff
@@ -0,0 +1,134 @@
+         this.frameBufferId = -1;
+         this.colorTextureId = -1;
+         this.depthBufferId = -1;
++
++        if (!useDepth && useStencil) {
++            var capabilities = org.lwjgl.opengl.GL.getCapabilities();
++            if (!capabilities.GL_ARB_texture_stencil8 && !capabilities.OpenGL44) {
++                throw new UnsupportedOperationException("Stencil-only buffers require GL_ARB_texture_stencil8 OR OpenGL 4.4");
++            }
++        }
```
Is there a use-case for this or is this going to be effectively dead code?

FiniteReality replied:
OpenGL technically supports it, and it's a supported combination of parameters, so rather than getting a confusing error about framebuffer completeness with no exception, I'd rather throw an exception beforehand.

You are correct though, in that the average user would probably not hit this branch

Shadows-of-Fire replied:
If this is perf-critical code I would suggest dropping this check.  Presumably the modder raising the incorrect flag combo would see a breakage elsewhere.

If not, we can keep it for informative purposes.

FiniteReality replied:
It's not performance critical, as it's only checked on construct. As far as I'm aware, vanilla prefers to reuse render targets rather than re-create them every frame.

shartte commented on patches/com/mojang/blaze3d/pipeline/RenderTarget.java.patch:
```diff
@@ -0,0 +1,134 @@
+         this.unbindRead();
+         this.unbindWrite();
++        if (this.stencilBufferId > -1 && this.stencilBufferId != this.depthBufferId) {
++            TextureUtil.releaseTextureId(this.stencilBufferId);;
```
Double semicolon, did this pass applyFormatting? 🤔

Matyrobbrt replied:
Spotless doesn't run on patches.

FiniteReality replied:
Whooops :sweat_smile:

FiniteReality committed:
Fix copy-paste fail in RenderTarget

FiniteReality committed:
Use an enum for the buffer type instead of two bools

FiniteReality commented on patches/net/minecraft/client/renderer/PostChain.java.patch:
```diff
@@ -0,0 +1,57 @@
+ import java.util.stream.Stream;
+ import javax.annotation.Nullable;
++
++import net.minecraft.client.renderer.PostPass.TargetInput.BufferType;
```
Dang. I *did* check this originally but then I renamed the class and IntelliJ "helped" :unamused:

FiniteReality committed:
Fix import whoopsie

FiniteReality committed:
Fix stenciling on the main target

FiniteReality committed:
Remove no-op method for enabling depth, pass desired width/height

FiniteReality commented on patches/net/minecraft/client/renderer/PostPass.java.patch:
```diff
@@ -0,0 +1,73 @@
+     @OnlyIn(Dist.CLIENT)
+-    public static record TargetInput(String samplerName, ResourceLocation targetId, boolean depthBuffer, boolean bilinear) implements PostPass.Input {
++    public static record TargetInput(String samplerName, ResourceLocation targetId, BufferType bufferType, boolean bilinear) implements PostPass.Input {
++        public enum BufferType {
```
Question: should this be `@OnlyIn` too, or should we move it to another type? This was the suggestion by @embeddedt to combine the separate bools into a single enum

FiniteReality replied:
Since nobody's responded to this in 3 weeks, I'm going to assume it's fine to leave as-is. If something breaks, we can fix it in a follow-up PR/commit.

FiniteReality committed:
Minimise patch by removing unnecessary indentation

Technici4n reviewed (changes requested):
I think that the overall strategy is sound, but I would remove all of the patches to make depth disable-able. That will already reduce the amount of patching by quite a bit. I would try to reduce the patch size even further. Finally, might be nice to reactivate `StencilEnableTest` to at least validate that enabling the stencil buffer won't cause everything to blow up.

XFactHD commented on patches/com/mojang/blaze3d/framegraph/FrameGraphBuilder.java.patch:
```diff
@@ -0,0 +1,46 @@
+         }
++
++        @Nullable
++        public abstract ResourceDescriptor<T> getDescriptor();
```
This method needs to not be abstract and instead return null by default.

XFactHD commented on patches/net/minecraft/client/renderer/PostChain.java.patch:
```diff
@@ -0,0 +1,48 @@
++                    useDepth |= renderDescriptor.useDepth();
++                    useStencil |= renderDescriptor.useStencil();
++            } else {
++                useDepth |= p_361871_.get(resourcelocation).get().useDepth;
```
Is there any reason to look this up three times instead of once?

FiniteReality committed:
Respond to XFact's feedback

Technici4n commented on patches/com/mojang/blaze3d/pipeline/MainTarget.java.patch:
```diff
@@ -0,0 +1,133 @@
+         GlStateManager._texParameter(3553, 10242, 33071);
+         GlStateManager._texParameter(3553, 10243, 33071);
+         GlStateManager._glFramebufferTexture2D(36160, 36064, 3553, this.colorTextureId, 0);
++        if (this.useDepth) {
```
Please remove all hooks to make `useDepth` disableable.

FiniteReality committed:
Remove stencil-only patches in MainTarget

Other patches are still necessary as other targets have that as a valid
combination in vanilla anyway, and post chains can be configured to render
to a target with depth disabled.

Technici4n commented on patches/com/mojang/blaze3d/pipeline/MainTarget.java.patch:
```diff
@@ -0,0 +1,87 @@
++    }
++
++    private MainTarget(net.neoforged.neoforge.client.event.ConfigureMainRenderTargetEvent e) {
++        super(e.useDepth(), e.useStencil());
```
`useDepth` shouldn't be in the event

FiniteReality committed:
Event cleanup and attempt other-platform support

- Changed the construction of MainTarget to use a hook instead
  - NeoForge now controls all parameters cleanly, with APIs that
    can be expanded for additional users (e.g. resolution scaling on
    Retina displays)
- Removed the use_stencil_buffer hook in PostChainConfig
  - This is unsupported on MacOS through most means:
    - Mac does not support ARB_texture_stencil8 (OpenGL 4.4)
    - Mac does not support ARB_texture_stenciling (OpenGL 4.3)
    - OpenGL 4.2 is the highest Mac supports
  - The code to wire this through remains for advanced users under the
    expectation they will perform the necessary tests themselves.
- Attempted to support Mac by falling back to a combined depth/stencil texture
  - This should be compatible in most cases as stencil-only buffers currently
    only exist for custom render targets.

Technici4n reviewed (changes requested):
You added even more complexity to support the stencil only case. I don't think this is the way to go. Stencil-only support should likely be moved to a separate PR so that we can focus on restoring previous behavior.

I would like more maintainers to review this PR and get their opinion. Specifically from @XFactHD whenever he can.

embeddedt commented:
@Technici4n I'm not really seeing much code that is specific to stencil-only mode, as opposed to being needed to support stenciling at all. Could you clarify the "even more complexity" you're referring to?

In general I haven't been following this PR *too* closely but this functionality has been missing, to my knowledge, since roughly 1.21.2 and it would be better to merge something that works than to keep bikeshedding it indefinitely.

Technici4n commented:
@embeddedt See https://github.com/neoforged/NeoForge/pull/1830 for a version that doesn't support stencil-only.
